[JSC] Allow direct load with offset from metadataTableRegister in baseline JIT resolve_scope and get_from_scope

ddegazio opened the pull request:
#### 588321aaad0462906bbadc73b0d9e633a89459f6
<pre>
[JSC] Allow direct load with offset from metadataTableRegister in baseline JIT resolve_scope and get_from_scope
<a href="https://bugs.webkit.org/show_bug.cgi?id=289289">https://bugs.webkit.org/show_bug.cgi?id=289289</a>
<a href="https://rdar.apple.com/146428148">rdar://146428148</a>

Reviewed by Yusuke Suzuki.

This patch adds a new common utility function computeGoodBaseAddressForMetadata to
the baseline JIT compiler, which either materializes the base address of a bytecode
instruction&apos;s metadata into a register, or attempts to offset from the metadata
table register directly if it can be encoded efficiently on the current platform.
This is used to remove some pointer arithmetic from the fast paths of resolve_scope
and get_from_scope when we don&apos;t have huge amounts of metadata.

Additionally, this patch refactors the implementation of put_to_scope to solve the
opposite issue - currently that function uses loadXFromMetadata() for all of its
metadata accesses, which at large offsets means we are redoing the pointer arithmetic
for each access. Now it uses the same new helper function to compute the address just
once.

* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITInlines.h:
(JSC::JIT::computeGoodBaseAddressForMetadata):
* Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
(JSC::JIT::emit_op_resolve_scope):
(JSC::JIT::emitSlow_op_resolve_scope):
(JSC::JIT::emit_op_get_from_scope):
(JSC::JIT::emitSlow_op_get_from_scope):
(JSC::JIT::emit_op_put_to_scope):

Canonical link: <a href="https://commits.webkit.org/291812@main">https://commits.webkit.org/291812@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f961d8efeb2d3371101fb4c3d1d8938398929df

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93921 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13506 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3241 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98927 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44448 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13803 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21937 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71667 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29019 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96923 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10245 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84850 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52003 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9927 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2492 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43764 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/86631 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80190 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2574 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100967 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/92587 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20973 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15290 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80682 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21225 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80816 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80032 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19980 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24581 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1945 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14130 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20957 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26135 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/115237 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20644 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24104 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22385 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/jit/JITInlines.h:
```diff
@@ -477,6 +477,44 @@ ALWAYS_INLINE ECMAMode JIT::ecmaMode<OpPutPrivateName>(OpPutPrivateName)
 }
 
+template<size_t minAlign, typename Bytecode>
+ALWAYS_INLINE MacroAssembler::Address JIT::computeGoodBaseAddressForMetadata(const Bytecode& bytecode, GPRReg metadataGPR)
```
Is there any value in calling this `computeGoodBaseAddressForMetadata`?  Why not just `computeBaseAddressForMetadata`?  When would we want a bad or less good base address?

ddegazio replied:
It's named that because it's responsible for deciding whether it would be good to use a `metadataTableRegister`-indirect load, but it is a mouthful, I'll change it.

MenloDorian commented on Source/JavaScriptCore/jit/JITInlines.h:
```diff
@@ -477,6 +477,44 @@ ALWAYS_INLINE ECMAMode JIT::ecmaMode<OpPutPrivateName>(OpPutPrivateName)
+    // implies that it is encodable at a larger access size.
+
+    using Metadata = typename Bytecode::Metadata;
+    static_assert(alignof(Metadata) >= minAlign);
```
I presume `alignof(Metadata)` have to be aligned on `minAlign` as well?  If so, this would be the better assert: `static_assert(WTF::roundUpToMultipleOf<minAlign>(alignof(Metadata)) == alignof(Metadata));`.

ddegazio replied:
Added in latest push.

MenloDorian commented on Source/JavaScriptCore/jit/JITInlines.h:
```diff
@@ -477,6 +477,44 @@ ALWAYS_INLINE ECMAMode JIT::ecmaMode<OpPutPrivateName>(OpPutPrivateName)
+    using Metadata = typename Bytecode::Metadata;
+    static_assert(alignof(Metadata) >= minAlign);
+    uint32_t metadataOffset = m_profiledCodeBlock->metadataTable()->offsetInMetadataTable(bytecode);
+    bool shouldEncodeMetadataOffsetInline = metadataOffset < 1 << 30;
```
Let's use `(1 << 30)` to be more explicit about this.

Are you sure that that `metadataOffset < 1 << 30` is the right condition for all non-ARM64 CPUs?  If you're not certain, the better thing would be to just disable this optimization for those CPUs (i.e. `shouldEncodeMetadataOffsetInline == false`), right?

ddegazio replied:
It may not strictly be ideal for every platform, although any all-or-nothing change is going to be imperfect. I think this is probably the right heuristic for x86_64 at least (since 32-bit displacements are fine to encode in a single instruction), to minimize friction it would probably be better to disable this on ARMv7/RISC-V since that'll pretty much just give the current behavior. I'll make that change in the next push.

MenloDorian commented on Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
```diff
@@ -960,37 +961,46 @@ void JIT::emit_op_resolve_scope(const JSInstruction* currentInstruction)
         // Inlined fast path for common types.
-        uint32_t metadataOffset = m_profiledCodeBlock->metadataTable()->offsetInMetadataTable(bytecode);
-        addPtr(TrustedImm32(metadataOffset), GPRInfo::metadataTableRegister, metadataGPR);
+        static_assert(!(Metadata::offsetOfResolveType() % 4));
```
I suggest using a `constexpr size_t minAlignment = 4;` here and below in this block instead of the literal 4.  That makes it explicit that they all need to be the same.

ddegazio replied:
Added in latest push.

MenloDorian commented on Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
```diff
@@ -960,37 +961,46 @@ void JIT::emit_op_resolve_scope(const JSInstruction* currentInstruction)
+        static_assert(!(Metadata::offsetOfGlobalLexicalBindingEpoch() % 4));
+        auto metadataAddress = computeGoodBaseAddressForMetadata<4>(bytecode, metadataGPR);
+
+        auto resolveTypeAddress = metadataAddress.withOffset(Metadata::offsetOfResolveType());
+        auto globalLexicalBindingEpochAddress = metadataAddress.withOffset(Metadata::offsetOfGlobalLexicalBindingEpoch());
```
How do we know that addresses with these offsets are encodable?  The `computeGoodBaseAddressForMetadata` above did not take into account this additional `Metadata::offsetOfResolveType()` nor `Metadata::offsetOfGlobalLexicalBindingEpoch()`.  Should `computeGoodBaseAddressForMetadata` take an range / addition offset allowance parameter to cover these?

ddegazio replied:
In `computeGoodBaseAddressForMetadata`, the offset we check is encodable is `metadataOffset + sizeof(Metadata)` - so, assuming correct alignment of each field (which the `static_assert`s around the callsite check), any field within the metadata object should be encodable too.

MenloDorian replied:
Oh, right.  I missed that.  You are right, `computeGoodBaseAddressForMetadata` already checks the worst case and guarantees that these offsets are reachable.

MenloDorian commented on Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
```diff
@@ -1535,6 +1569,18 @@ void JIT::emit_op_put_to_scope(const JSInstruction* currentInstruction)
+    static_assert(!(Metadata::offsetOfOperand() % 8));
+    static_assert(!(Metadata::offsetOfWatchpointSet() % 8));
+    auto metadataAddress = computeGoodBaseAddressForMetadata<8>(bytecode, metadataGPR);
+    auto getPutInfoAddress = metadataAddress.withOffset(Metadata::offsetOfGetPutInfo());
+    auto structureAddress = metadataAddress.withOffset(Metadata::offsetOfStructure());
+    auto operandAddress = metadataAddress.withOffset(Metadata::offsetOfOperand());
+    auto watchpointSetAddress = metadataAddress.withOffset(Metadata::offsetOfWatchpointSet());
```
Ditto, how do we know that these Addresses will produce encodable offsets?

MenloDorian commented on Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
```diff
@@ -1535,6 +1569,18 @@ void JIT::emit_op_put_to_scope(const JSInstruction* currentInstruction)
+    static_assert(!(Metadata::offsetOfGetPutInfo() % 8));
+    static_assert(!(Metadata::offsetOfStructure() % 8));
+    static_assert(!(Metadata::offsetOfOperand() % 8));
+    static_assert(!(Metadata::offsetOfWatchpointSet() % 8));
```
Ditto: let's use a `constexpr size_t minAlignment = 8;` here instead of these literal 8s.

ddegazio replied:
Added in latest push - actually based on `alignof(void*)` since these are pointers and not 64-bit values.

ddegazio committed:
[JSC] Allow direct load with offset from metadataTableRegister in baseline JIT resolve_scope and get_from_scope
https://bugs.webkit.org/show_bug.cgi?id=289289
rdar://146428148

Reviewed by Yusuke Suzuki.

This patch adds a new common utility function computeGoodBaseAddressForMetadata to
the baseline JIT compiler, which either materializes the base address of a bytecode
instruction's metadata into a register, or attempts to offset from the metadata
table register directly if it can be encoded efficiently on the current platform.
This is used to remove some pointer arithmetic from the fast paths of resolve_scope
and get_from_scope when we don't have huge amounts of metadata.

Additionally, this patch refactors the implementation of put_to_scope to solve the
opposite issue - currently that function uses loadXFromMetadata() for all of its
metadata accesses, which at large offsets means we are redoing the pointer arithmetic
for each access. Now it uses the same new helper function to compute the address just
once.

* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITInlines.h:
(JSC::JIT::computeGoodBaseAddressForMetadata):
* Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
(JSC::JIT::emit_op_resolve_scope):
(JSC::JIT::emitSlow_op_resolve_scope):
(JSC::JIT::emit_op_get_from_scope):
(JSC::JIT::emitSlow_op_get_from_scope):
(JSC::JIT::emit_op_put_to_scope):

Canonical link: https://commits.webkit.org/291812@main
